jsk_pcl_ros/src/octomap_server_contact_nodelet.cpp

```diff
@@ -437,63 +454,200 @@ namespace jsk_pcl_ros
+      unknownNodesVis.markers[0].scale.z = size;
+      unknownNodesVis.markers[0].color = m_colorUnknown;
+
+      if (unknownNodesVis.markers[0].points.size() > 0) {
+        unknownNodesVis.markers[0].action = visualization_msgs::Marker::ADD;
+      }
+      else {
+        unknownNodesVis.markers[0].action = visualization_msgs::Marker::DELETE;
+      }
+      m_umarkerPub.publish(unknownNodesVis);
+
+      // publish unknown grid as pointcloud
+      sensor_msgs::PointCloud2 unknownRosCloud;
+      pcl::toROSMsg (unknownCloud, unknownRosCloud);
+      unknownRosCloud.header.frame_id = m_worldFrameId;
+      unknownRosCloud.header.stamp = rostime;
+      m_unknownPointCloudPub.publish(unknownRosCloud);
+
+      // publish frontier grid as marker
+      if (publishFrontierMarkerArray) {
+        visualization_msgs::MarkerArray frontierNodesVis;
+        frontierNodesVis.markers.resize(1);
+        pcl::PointCloud<pcl::PointXYZ> frontierCloud;
+        double resolution = m_octreeContact->getResolution();
+        // how many resolution-size grids are in one edge
+        int x_num = int(((m_occupancyMaxX - m_occupancyMinX) / resolution));
+        int y_num = int(((m_occupancyMaxY - m_occupancyMinY) / resolution));
+        int z_num = int(((m_occupancyMaxZ - m_occupancyMinZ) / resolution));
+        std::vector< std::vector< std::vector<int> > > check_unknown(x_num, std::vector< std::vector<int> >(y_num, std::vector<int>(z_num)));
+        for (int i=0; i<x_num; i++) {
+          for (int j=0; j<y_num; j++) {
+            for (int k=0; k<z_num; k++) {
+              check_unknown[i][j][k] = 0;
+            }
+          }
+        }
+        std::vector< std::vector< std::vector<int> > > check_occupied(x_num, std::vector< std::vector<int> >(y_num, std::vector<int>(z_num)));
+        for (int i=0; i<x_num; i++) {
+          for (int j=0; j<y_num; j++) {
+            for (int k=0; k<z_num; k++) {
+              check_occupied[i][j][k] = 0;
+            }
+          }
+        }
+        std::vector< std::vector< std::vector<int> > > check_frontier(x_num, std::vector< std::vector<int> >(y_num, std::vector<int>(z_num)));
+        for (int i=0; i<x_num; i++) {
+          for (int j=0; j<y_num; j++) {
+            for (int k=0; k<z_num; k++) {
+              check_frontier[i][j][k] = 0;
+            }
+          }
+        }
+
+        // for all unknown grids, store its information to array
+        for (point3d_list::iterator it_unknown = unknownLeaves.begin();
+             it_unknown != unknownLeaves.end();
+             it_unknown++) {
+          // get center of unknown grids
+          double x_unknown = it_unknown->x();
+          double y_unknown = it_unknown->y();
+          double z_unknown = it_unknown->z();
+          check_unknown[int(std::round((x_unknown - m_occupancyMinX) / resolution - 1))][int(std::round((y_unknown - m_occupancyMinY) / resolution - 1))][int(std::round((z_unknown - m_occupancyMinZ) / resolution - 1))] = 1;
```

Comment:
too long. set new variable for `int(std::round((x_unknown - m_occupancyMinX) / resolution - 1))` and others